testsuite/model/src/test/java/org/keycloak/testsuite/model/infinispan/CacheExpirationTest.java

```diff
@@ -0,0 +1,137 @@
+import org.hamcrest.Matchers;
+import org.infinispan.Cache;
+import org.junit.Assume;
+import org.junit.Test;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.junit.Assume.assumeThat;
+
+
+/**
+ *
+ * @author hmlnarik
+ */
+@RequireProvider(InfinispanConnectionProvider.class)
+public class CacheExpirationTest extends KeycloakModelTest {
+
+    @Test
+    public void testCacheExpiration() throws Exception {
+        AtomicLong putTime = new AtomicLong();
+        inComittedTransaction(session -> {
+            InfinispanConnectionProvider provider = session.getProvider(InfinispanConnectionProvider.class);
+            Cache<String, Object> cache = provider.getCache(InfinispanConnectionProvider.WORK_CACHE_NAME);
+            cache.entrySet().stream()
+              .filter(me -> me.getValue() instanceof AuthenticationSessionAuthNoteUpdateEvent)
+              .forEach((c, me) -> c.remove(me.getKey()));
+
+            putTime.set(System.currentTimeMillis());
+            cache.put("1-2", AuthenticationSessionAuthNoteUpdateEvent.create("g1", "p1", "r1", Collections.emptyMap()), 20000, TimeUnit.MILLISECONDS);
+            cache.put("1-2-3", AuthenticationSessionAuthNoteUpdateEvent.create("g2", "p2", "r2", Collections.emptyMap()), 20000, TimeUnit.MILLISECONDS);
+        });
+
+        assumeThat("jmap output format unsupported", getNumberOfInstancesOfClass(AuthenticationSessionAuthNoteUpdateEvent.class), notNullValue());
+
+        assertThat(getNumberOfInstancesOfClass(AuthenticationSessionAuthNoteUpdateEvent.class), is(2));
+
+        AtomicInteger maxCountOfInstances = new AtomicInteger();
+        AtomicInteger minCountOfInstances = new AtomicInteger(100);
+        inIndependentFactories(4, 5 * 60, () -> {
+            log.debug("Joining the cluster");
+            inComittedTransaction(session -> {
+                InfinispanConnectionProvider provider = session.getProvider(InfinispanConnectionProvider.class);
+                Cache<String, Object> cache = provider.getCache(InfinispanConnectionProvider.WORK_CACHE_NAME);
+                do {
+                    try { Thread.sleep(1000); } catch (InterruptedException ex) {}
+                } while (! cache.getAdvancedCache().getDistributionManager().isJoinComplete());
+                cache.keySet().forEach(s -> {});
+            });
+            log.debug("Cluster joined");
+            int c = getNumberOfInstancesOfClass(AuthenticationSessionAuthNoteUpdateEvent.class);
+            maxCountOfInstances.getAndAccumulate(c, Integer::max);
+            assumeThat("Seems we're running on a way too slow a computer", System.currentTimeMillis() - putTime.get(), Matchers.lessThan(20000L));
+
+            // Wait for at most 3 minutes which is much more than 15 seconds expiration set in DefaultInfinispanConnectionProviderFactory
+            for (int i = 0; i < 3 * 60; i++) {
```

Comment:
Would it make sense to do a gc in here to make this happen quicker?

Reply:
It would, and is either already done or logic that would be used for GCing is involved by the `jmap -histo:live` command that is invoked in `getNumberOfInstancesOfClass` (the `:live` part is responsible for doing this)